### PR TITLE
Fix crash in native crypto on concurrent cipher object use with GCM

### PR DESCRIPTION
**Pull request by ZainabF92**
Porting PR [openj9-openjdk-jdk16#66](https://github.com/ibmruntimes/openj9-openjdk-jdk16/pull/66)

Signed-off-by: Zainab Fatmi <zainab@ibm.com>

### CODE REVIEW

**Comment by ashbm5** [comment]
FYI @pshipton @keithc-ca

File: closed/src/java.base/share/classes/com/sun/crypto/provider/NativeGaloisCounterMode.java [hunk 1]
```diff
@@ -326,40 +328,51 @@ int encrypt(byte[] in, int inOfs, int len, byte[] out, int outOfs) {
      */
     int encryptFinal(byte[] in, int inOfs, int len, byte[] out, int outOfs)
         throws IllegalBlockSizeException, ShortBufferException {
+        byte[] keyCopy;
```
Comment (pshipton) [inline closed/src/java.base/share/classes/com/sun/crypto/provider/NativeGaloisCounterMode.java:331]: The first three don't need to be declared out of the scope where they are used.

Comment (ZainabF92) [reply closed/src/java.base/share/classes/com/sun/crypto/provider/NativeGaloisCounterMode.java:331]: fixed

File: closed/src/java.base/share/classes/com/sun/crypto/provider/NativeGaloisCounterMode.java [hunk 2]
```diff
@@ -416,65 +431,75 @@ int decryptFinal(byte[] in, int inOfs, int len,
...
         throws IllegalBlockSizeException, AEADBadTagException,
         ShortBufferException {
-        if (len < tagLenBytes) {
-            throw new AEADBadTagException("Input too short - need tag");
-        }
-
-        // do this check here can also catch the potential integer overflow
-        // scenario for the subsequent output buffer capacity check.
-        checkDataLength(ibuffer.size(), (len - tagLenBytes));
+        byte[] keyCopy;
```
[the remark this answers is not in the document]
Comment (ZainabF92) [reply closed/src/java.base/share/classes/com/sun/crypto/provider/NativeGaloisCounterMode.java:434]: fixed

File: closed/src/java.base/share/classes/com/sun/crypto/provider/NativeGaloisCounterMode.java [hunk 3]
```diff
@@ -326,40 +328,51 @@ int encrypt(byte[] in, int inOfs, int len, byte[] out, int outOfs) {
...
-        byte[] aad = ((aadBuffer == null || aadBuffer.size() == 0) ? emptyAAD : aadBuffer.toByteArray());
+            aad = (((aadBuffer == null) || (aadBuffer.size() == 0)) ? emptyAAD : aadBuffer.toByteArray());
 
-        int ret = nativeCrypto.GCMEncrypt(key, key.length,
-                iv, iv.length,
-                in, inOfs, len,
-                out, outOfs,
-                aad, aad.length, tagLenBytes);
+            // avoid any race conditions in OpenSSL when the key/iv are used concurrently
+            keyCopy = key.clone();
```
Comment (pshipton) [inline closed/src/java.base/share/classes/com/sun/crypto/provider/NativeGaloisCounterMode.java:363]: The clones seem unnecessary, key and iv are already cloned in init(). GCMEncrypt and GCMDecrypt are synchronized now, so they can't run at the same time. I don't see how key/iv can be used concurrently.

Comment (ZainabF92) [reply closed/src/java.base/share/classes/com/sun/crypto/provider/NativeGaloisCounterMode.java]: yeah, that makes sense

**Comment by pshipton** [comment]
jenkins test sanity.openjdk alinux jdk11

**Comment by pshipton** [comment]
jenkins test sanity.openjdk alinux64 jdk11

**Review by pshipton — APPROVED** [review]

**Pull request merged**